src/server/typingsInstaller/nodeTypingsInstaller.ts

```diff
@@ -106,40 +89,69 @@ namespace ts.server.typingsInstaller {
             }
         }
 
-        protected runTsd(cachePath: string, typingsToInstall: string[], postInstallAction: (installedTypings: string[]) => void): void {
-            const id = this.tsdRunCount;
-            this.tsdRunCount++;
-            const tsdPath = combinePaths(this.npmBinPath, "tsd");
-            const command = `${tsdPath} install ${typingsToInstall.join(" ")} -ros`;
-            if (this.log.isEnabled()) {
-                this.log.writeLine(`Running tsd ${id}, command '${command}'. cache path '${cachePath}'`);
-            }
-            this.exec(command, { cwd: cachePath }, (err, stdout, stderr) => {
+        protected runInstall(cachePath: string, typingsToInstall: string[], postInstallAction: (installedTypings: string[]) => void): void {

```

Comment:
I would probably write it a bit differently to reduce nesting and code duplication, smth like below (warning, typed directly in browser, may contain ~~peanuts~~ errors)

```ts
private execAsync(prefix: string, command: string, cwd: string, requestId: number, cb: (err: Error, stdout: string, stderr: string) => void) {
    if (this.log.isEnabled()) {
        this.log.writeLine(`#${requestId} running command '${command}'.`);
    }
    this.exec(command, { cwd }, (err, stdout, stderr) => {
        if (this.log.isEnabled()) {
            this.log.writeLine(`${prefix} #${requestId} stdout: ${stdout}`);
            this.log.writeLine(`${prefix} #${requestId} stderr: ${stderr}`);
        }
        cb(err, stdout, stderr);
    });
}

protected runInstall(cachePath: string, typingsToInstall: string[], postInstallAction: (installedTypings: string[]) => void): void {
    const id = this.installRunCount;
    this.installRunCount++;
    let execInstallCmdCount = 0;
    let filteredTypings: string[] = [];
    
    for (const typing of typingsToInstall) {
        const command = `npm view @types/${typing} --silent name`;
        this.execAsync("npm view @types", command, cachePath, id, (err, stdout, stderr) => {
            if (stdout) {
                filteredTypings.push(stdout);
            }
            execInstallCmdCount++;
            if (execInstallCmdCount === typingsToInstall.length) {
                installFilteredTypings(this, filteredTypings);
            }
        })
    }

    function installFilteredTypings(self: NodeTypingsInstaller, filteredTypings: string[]) {
        const command = `npm install ${filteredTypings.map(t => "@types/" + t).join(" ")} --save-dev`;
        self.execAsync("npm install", command, cachePath, id, (err, stdout, stderr) => {
            if (stdout) {
                reportInstalledTypings(self)
            }
        })
    }

    function reportInstalledTypings(self: NodeTypingsInstaller) {
        const command = "npm ls -json";
        self.execAsync("npm ls", command, cachePath, id, (err, stdout, stderr) => {
            let installedTypings: string[];
            try {
                const response = JSON.parse(stdout);
                if (response.dependencies) {
                    installedTypings = getOwnKeys(response.dependencies);
                }
            }
            catch (e) {
                self.log.writeLine(`Error parsing installed @types dependencies. Error details: ${e.message}`);
            }
            postInstallAction(installedTypings || []);
        });
    }
```